File: src/main/java/com/shadowbattler/simulator/model/Move.java
package com.shadowbattler.simulator.model;

import com.fasterxml.jackson.annotation.JsonIgnoreProperties;
import com.fasterxml.jackson.databind.annotation.JsonDeserialize;
import com.fasterxml.jackson.databind.util.StdConverter;

@JsonDeserialize(converter = MoveJsonConverter.class)
public record Move(
    String moveId,
    String name,
    String abbreviation,
    Type type,
    int power,
    int energy,
    int energyGain,
    Stats3<Integer> buffsSelf,
    Stats3<Integer> buffsOpponent,
    double buffApplyChance,
    String archetype,
    int turns
) {}

class MoveJsonConverter extends StdConverter<MoveJsonConverter.MoveJson, Move> {
    @JsonIgnoreProperties(ignoreUnknown=true)
    static class MoveJson {
        public String moveId;
        public String name;
        public String abbreviation;
        public Type type;
        public int power;
        public int energy;
        public int energyGain;
        public int cooldown;
        public int[] buffs;
        public int[] buffsSelf;
        public int[] buffsOpponent;
        public String buffTarget;
        public double buffApplyChance;
        public String archetype;
        public int turns;
    }

    @Override
    public Move convert(MoveJson moveJson) {
        int[] buffsSelfArr = moveJson.buffsSelf;
        int[] buffsOpponentArr = moveJson.buffsOpponent;

        if (moveJson.buffTarget != null) {
            if (moveJson.buffTarget.equalsIgnoreCase("self") && buffsSelfArr == null) {
                buffsSelfArr = moveJson.buffs;
            } else if (moveJson.buffTarget.equalsIgnoreCase("opponent") && buffsOpponentArr == null) {
                buffsOpponentArr = moveJson.buffs;
            }
        }

        final Stats3<Integer> buffsSelfStats3;
        if (buffsSelfArr != null && buffsSelfArr.length >= 2) {
            buffsSelfStats3 = new Stats3<>(buffsSelfArr[0], buffsSelfArr[1], 0);
        } else {
            buffsSelfStats3 = null;
        }

        final Stats3<Integer> buffsOpponentStats3;
        if (buffsOpponentArr != null && buffsOpponentArr.length >= 2) {
            buffsOpponentStats3 = new Stats3<>(buffsOpponentArr[0], buffsOpponentArr[1], 0);
        } else {
            buffsOpponentStats3 = null;
        }

        return new Move(
            moveJson.moveId, 
            moveJson.name, 
            moveJson.abbreviation, 
            moveJson.type, 
            moveJson.power, 
            moveJson.energy, 
            moveJson.energyGain, 
            buffsSelfStats3, 
            buffsOpponentStats3, 
            moveJson.buffApplyChance, 
            moveJson.archetype,
            moveJson.turns
        );
    }
}

File: frontend/types/Move.ts
import { Species } from "./Species"
import { Stats3 } from "./Stats3"
import { Type } from "./Type"

export type Move = {
    moveId: string,
    name: string,
    abbreviation: string,
    type: Type,
    power: number,
    energy: number,
    energyGain: number,
    buffsSelf: Stats3<number>,
    buffsOpponent: Stats3<number>,
    buffApplyChance: number,
    archetype: string,
    turns: number
}

export const Move = {
    eliteSymbol: '*',
    legacySymbol: '†',

    getSymbol(move: Move, species: Species) {
        if (species.eliteMoves?.includes(move.moveId)) {
            return this.eliteSymbol;
        } else if (species.legacyMoves?.includes(move.moveId)) {
            return this.legacySymbol;
        }
        return '';
    },

    getNameWithSymbols(move: Move, species?: Species) {
        return species ? move.name + this.getSymbol(move, species) : move.name;
    }
}